### PR TITLE
[Feature] Federation aware connections

### PR DESCRIPTION
**Pull request by typfel**
## What's new in this PR?

This is part of a series of PRs which refactor connections and make them federation aware.

- `ConnectionRequestStrategy` replaces `ZMConnectionTranscoder`
- ZMConnection are no longer inserted locally in UI context to be synchronised by SE context at a later point. This behaviour is now implemented  by the `ConnectToUserAction` which means that `ZMConnection` objects are only inserted in SE context which avoid many possible race conditions.
- ZMConnection no longer synchronises any state by observing the `status` field. This behaviour is now implemented by the `UpdateConnectionAction`.

## Dependencies

Needs releases with:

- [x] https://github.com/wireapp/wire-ios-data-model/pull/1241

### CODE REVIEW

File: Sources/Request Strategies/Connection/ConnectionRequestStrategy.swift [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+
+                    if !connectionList.hasMore {
+                        self?.syncProgress.finishCurrentSyncPhase(phase: .fetchingConnections)
+                        self?.isFetchingAllConnections = false
+                    }
+                case .failure:
+                    self?.syncProgress.failCurrentSyncPhase(phase: .fetchingConnections)
+                }
+            }
+        }
```
Comment (David-Henner) [inline Sources/Request Strategies/Connection/ConnectionRequestStrategy.swift:129]: the two fetch completion blocks look identical, is there a way to reduce the duplication here ?

File: Sources/Request Strategies/Connection/ConnectionRequestStrategy.swift [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+        } else {
+            if let userID = object.to.remoteIdentifier {
+                connectionByIDSync.sync(identifiers: Set(arrayLiteral: userID))
+            }
+        }
+    }
+
+    func cancel(_ object: ZMConnection) {
+        
+    }
```
Comment (David-Henner) [inline Sources/Request Strategies/Connection/ConnectionRequestStrategy.swift:167]: is implementation missing?

File: Sources/Request Strategies/Connection/ConnectionRequestStrategy.swift [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+                conversationEvent?.process(in: managedObjectContext)
+            default:
+                break
+            }
+        }
+    }
+
+}
+
+class ConnectionByIDTranscoder: IdentifierObjectSyncTranscoder {
```
Comment (David-Henner) [inline Sources/Request Strategies/Connection/ConnectionRequestStrategy.swift:195]: will this `ConnectionByIDTranscoder` be completely replaced by `ConnectionByQualifiedIDTranscoder` at some point?

File: Sources/Payloads/Processing/Payload+Processing.swift [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+                                                                          selfClient: selfClient)
+                } else {
+                    missingClient.markClientAsInvalidAfterFailingToRetrievePrekey(selfClient: selfClient)
+                }
+
+
+            }
+        }
+
+        let hasMoreMissingClients = (selfClient.missingClients?.count ?? 0) > 0
```
Comment (billypchan) [inline Sources/Payloads/Processing/Payload+Processing.swift:100]:
```suggestion
        let hasMoreMissingClients = selfClient.missingClients?.isEmpty == false
```

**Pull request merged**